puppet/lib/puppet/indirector/facts/puppetdb.rb

```diff
@@ -64,24 +64,21 @@ def find(request)
           profile("Parse fact query response (size: #{response.body.size})",
                   [:puppetdb, :facts, :find, :parse_response, request.key]) do
             result = JSON.parse(response.body)
-            # Note: the Inventory Service API appears to expect us to return nil here
-            # if the node isn't found.  However, PuppetDB returns an empty array in
-            # this case; for now we will just look for that condition and assume that
-            # it means that the node wasn't found, so we will return nil.  In the
-            # future we may want to improve the logic such that we can distinguish
-            # between the "node not found" and the "no facts for this node" cases.
-            if result.empty?
-              return nil
-            end
+
             facts = result.inject({}) do |a,h|
               a.merge(h['name'] => h['value'])
             end
+
             Puppet::Node::Facts.new(request.key, facts)
           end
         else
           # Newline characters cause an HTTP error, so strip them
           raise "[#{response.code} #{response.message}] #{response.body.gsub(/[\r\n]/, '')}"
         end
+      rescue NotFoundError => e
```

Comment:
And now we catch this exception instead of looking for an empty result, as we get a 404 and that is surfaced via exception from the http wrapper.